Avoid undesirable allocation when device is associated with multiple …

kikimo opened the pull request:
…NUMA Nodes

suppose there are two devices dev1 and dev2, each has NUMA Nodes associated as below:
  dev1: numa1
  dev2: numa1, numa2

and we request a device from numa2, currently filterByAffinity() will return
[], [dev1, dev2], [] if loop of available devices produce a sequence of [dev1, dev2],
that is is not desirable as what we truely expect is an allocation of dev2 from numa2.

<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?
/kind bug

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature
/kind design

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:
Avoid undesirable allocation when device is associated with multiple NUMA Nodes.

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

kikimo commented:
> Thanks for your pull request. Before we can look at your pull request, you'll need to sign a Contributor License Agreement (CLA).
> 
> 📝 **Please follow instructions at https://git.k8s.io/community/CLA.md#the-contributor-license-agreement to sign the CLA.**
> 
> It may take a couple minutes for the CLA signature to be fully registered; after that, please reply here with a new comment and we'll verify. Thanks.
> 
> * If you've already signed a CLA, it's possible we don't have your GitHub username or you're using a different email address.  Check your existing CLA data and verify that your [email is set on your git commits](https://help.github.com/articles/setting-your-email-in-git/).
> * If you signed the CLA as a corporation, please sign in with your organization's credentials at https://identity.linuxfoundation.org/projects/cncf to be authorized.
> * If you have done the above and are still having issues with the CLA being reported as unsigned, please log a ticket with the Linux Foundation Helpdesk: https://support.linuxfoundation.org/
> * Should you encounter any issues with the Linux Foundation Helpdesk, send a message to the backup e-mail support address at: [login-issues@jira.linuxfoundation.org](mailto:login-issues@jira.linuxfoundation.org)

signed

fromanirh commented:
/ok-to-test

SergeyKanzhelev commented:
/assign @fromanirh 
/triage accepted
/priority important-soon

please comment whether it's a bug or just a cleanup

kikimo commented:
> /assign @fromanirh
> /triage accepted
> /priority important-soon
> 
> please comment whether it's a bug or just a cleanup

This is obviously a bug. It's a problem we found in our production cluster. When triggered topology manager with single-numa-policy might return devices spanning serveral NUMA Nodes.

fromanirh commented:
/kind bug

fromanirh commented on pkg/kubelet/cm/devicemanager/manager.go:
```diff
@@ -808,21 +808,31 @@ func (m *ManagerImpl) filterByAffinity(podUID, contName, resource string, availa
+		// only when it found that this node meets the affinity requirement and
+		// set fromAffinity as target list. Otherwise it will keep looping and:
+		//    i: set nodeWithoutTopology as target list when targetList is nil and
+		//   ii: set notFromAffinity as target list when target is nil or nodeWithoutTopology
```
so, if we have devices declaring affinity to more than a NUMA node, we should prefer devices which are local only to the node who have the hint for, right?
IOW, we should prefer devices whose affinity is more restrictive, right? in this example, dev1 has affinity=1 numa nodes while dev2 has affinity=2 numa node.

fromanirh commented on pkg/kubelet/cm/devicemanager/manager.go:
```diff
@@ -808,21 +808,31 @@ func (m *ManagerImpl) filterByAffinity(podUID, contName, resource string, availa
+		// set fromAffinity as target list. Otherwise it will keep looping and:
+		//    i: set nodeWithoutTopology as target list when targetList is nil and
+		//   ii: set notFromAffinity as target list when target is nil or nodeWithoutTopology
+		var targetList *[]string
```
I'd prefer to avoid this indirection in the code

kikimo replied:
@fromanirh I think a better way to fix this problem is to sort nodes list used in the loop in the order that nodes with numa affinity comes first then nodes without numa affinity and nodes without topology come as last, then we  no longer need the indirection list. If this is ok for you, I'll take another try.

fromanirh replied:
This seems a bit better, yes.

fromanirh commented on pkg/kubelet/cm/devicemanager/manager.go:
```diff
@@ -808,21 +808,31 @@ func (m *ManagerImpl) filterByAffinity(podUID, contName, resource string, availa
 		}
+
+		if targetList != nil {
+			*targetList = append(*targetList, d)
```
Do we have unit tests which exercise this flow and highlight the bug? If not, let's add them

ffromani reviewed (commented):
Yep, this seems like a bug

klueska commented:
Agreed. This is definitely a bug. Thanks for finding this. Let’s find a cleaner way of implementing the fix though.

fromanirh commented on pkg/kubelet/cm/devicemanager/manager.go:
```diff
@@ -808,21 +808,31 @@ func (m *ManagerImpl) filterByAffinity(podUID, contName, resource string, availa
+		// into this list. For NUMA node that has the device, this loop will break
+		// only when it found that this node meets the affinity requirement and
+		// set fromAffinity as target list. Otherwise it will keep looping and:
+		//    i: set nodeWithoutTopology as target list when targetList is nil and
```
It seems a comment of mine got lost. Let me retry.
IIUC, the problem is that we should de-prioritize devices who are affine to more than one numa node. When filtering devices, we should prefer devices who are compatible with the hint (obviously), but which have the stricter NUMA affinity set.
IOW, having hint pointing to NUMACELL2 between dev1 affine to NUMACELL2 (affinity=1 numa cell) and dev2 affine to NUMACELL1 and NUMACELL2 (affinity= 2 numa cells), prefer dev1.
Is this a good summarization?

klueska commented:
/assign @klueska

kikimo commented:
@freeformz @klueska I just push another patch with unit test, please take a look

kikimo commented:
/retest

kikimo committed:
Avoid undesirable allocation when device is associated with multiple NUMA Nodes

suppose there are two devices dev1 and dev2, each has NUMA Nodes associated as below:
  dev1: numa1
  dev2: numa1, numa2

and we request a device from numa2, currently filterByAffinity() will return
[], [dev1, dev2], [] if loop of available devices produce a sequence of [dev1, dev2],
that is is not desirable as what we truely expect is an allocation of dev2 from numa2.

kikimo reviewed (commented):
right

kikimo commented:
@freeformz @klueska found that I forgot to include the unit test codes, just push another commit to submit them, please take another look

kikimo commented:
> @kikimo: The following test **failed**, say `/retest` to rerun all failed tests:
> 
> Test name	Commit	Details	Rerun command
> pull-kubernetes-e2e-kind-ipv6	[94ca91d](https://github.com/kubernetes/kubernetes/commit/94ca91d058441149ef6a6d1c27b3b2be67fe95a3)	[link](https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/101893/pull-kubernetes-e2e-kind-ipv6/1393952162484064256)	`/test pull-kubernetes-e2e-kind-ipv6`
> [Full PR test history](https://prow.k8s.io/pr-history?org=kubernetes&repo=kubernetes&pr=101893). [Your PR dashboard](https://prow.k8s.io/pr?query=is%3Apr%20state%3Aopen%20author%3Akikimo). Please help us cut down on flakes by [linking to](https://git.k8s.io/community/contributors/devel/sig-testing/flaky-tests.md#filing-issues-for-flaky-tests) an [open issue](https://github.com/kubernetes/kubernetes/issues?q=is:issue+is:open) when you hit one in your PR.

/retest

kikimo commented:
> @kikimo: The following tests **failed**, say `/retest` to rerun all failed tests:
> 
> Test name	Commit	Details	Rerun command
> pull-kubernetes-e2e-kind-ipv6	[94ca91d](https://github.com/kubernetes/kubernetes/commit/94ca91d058441149ef6a6d1c27b3b2be67fe95a3)	[link](https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/101893/pull-kubernetes-e2e-kind-ipv6/1393952162484064256)	`/test pull-kubernetes-e2e-kind-ipv6`
> pull-kubernetes-e2e-gce-ubuntu-containerd	[94ca91d](https://github.com/kubernetes/kubernetes/commit/94ca91d058441149ef6a6d1c27b3b2be67fe95a3)	[link](https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/101893/pull-kubernetes-e2e-gce-ubuntu-containerd/1393952162412761088/)	`/test pull-kubernetes-e2e-gce-ubuntu-containerd`
> [Full PR test history](https://prow.k8s.io/pr-history?org=kubernetes&repo=kubernetes&pr=101893). [Your PR dashboard](https://prow.k8s.io/pr?query=is%3Apr%20state%3Aopen%20author%3Akikimo). Please help us cut down on flakes by [linking to](https://git.k8s.io/community/contributors/devel/sig-testing/flaky-tests.md#filing-issues-for-flaky-tests) an [open issue](https://github.com/kubernetes/kubernetes/issues?q=is:issue+is:open) when you hit one in your PR.

/retest

kikimo commented:
> @kikimo: The following test **failed**, say `/retest` to rerun all failed tests:
> 
> Test name	Commit	Details	Rerun command
> pull-kubernetes-e2e-gce-ubuntu-containerd	[94ca91d](https://github.com/kubernetes/kubernetes/commit/94ca91d058441149ef6a6d1c27b3b2be67fe95a3)	[link](https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/101893/pull-kubernetes-e2e-gce-ubuntu-containerd/1393967452592803840/)	`/test pull-kubernetes-e2e-gce-ubuntu-containerd`
> [Full PR test history](https://prow.k8s.io/pr-history?org=kubernetes&repo=kubernetes&pr=101893). [Your PR dashboard](https://prow.k8s.io/pr?query=is%3Apr%20state%3Aopen%20author%3Akikimo). Please help us cut down on flakes by [linking to](https://git.k8s.io/community/contributors/devel/sig-testing/flaky-tests.md#filing-issues-for-flaky-tests) an [open issue](https://github.com/kubernetes/kubernetes/issues?q=is:issue+is:open) when you hit one in your PR.

/retest

kikimo commented:
> @freeformz @klueska found that I forgot to include the unit test codes, just push another commit to submit them, please take another look

@fromanirh  any update?

fromanirh commented:
> > @freeformz @klueska found that I forgot to include the unit test codes, just push another commit to submit them, please take another look
> 
> @fromanirh any update?

I will review today (May 18, 2021), worst case tomorrow.

klueska commented on pkg/kubelet/cm/devicemanager/manager.go:
```diff
@@ -791,9 +791,23 @@ func (m *ManagerImpl) filterByAffinity(podUID, contName, resource string, availa
 
-	// Sort the list of nodes by how many devices they contain.
+	// Sort the list of nodes by nodes from affinity set, nodes from none-affinity
+	// set, nodes without topology set and how many devices they contain.
 	sort.Slice(nodes, func(i, j int) bool {
-		return perNodeDevices[i].Len() < perNodeDevices[j].Len()
+		ni, nj := nodes[i], nodes[j]
+		// 1. node[i] from to affninity set
+		if hint.NUMANodeAffinity.IsSet(ni) {
+			return !hint.NUMANodeAffinity.IsSet(nj) || perNodeDevices[ni].Len() < perNodeDevices[nj].Len()
+		}
+
+		// 2. node[i] from none-affinity set
+		if ni != nodeWithoutTopology {
+			return !hint.NUMANodeAffinity.IsSet(nj) &&
+				(nj == nodeWithoutTopology || perNodeDevices[ni].Len() < perNodeDevices[nj].Len())
+		}
+
+		// 3. node[i] from node without topology set
+		return nj == nodeWithoutTopology && perNodeDevices[ni].Len() < perNodeDevices[nj].Len()
```
In general, this is a much better approach than what was being proposed before. And I *think* this looks correct, but its still a little complicated to verify for my tastes. I will suggest an alternate (but hopefully clearer) way of doing the same later today.

kikimo replied:
@klueska just update the sorting algorithm, PTAL

klueska commented on pkg/kubelet/cm/devicemanager/manager_test.go:
```diff
@@ -669,6 +670,106 @@ type TestResource struct {
+	}
+}
+
+type fakeTopologyManager struct {
+	topologymanager.Manager
+}
+
+func newFakeTopologyManager() topologymanager.Manager {
+	return &fakeTopologyManager{}
+}
+
+func (m *fakeTopologyManager) GetAffinity(podUID string, containerName string) topologymanager.TopologyHint {
+	// return hint with numa2 set
+	affinity, _ := bitmask.NewBitMask(2)
+	return topologymanager.TopologyHint{
+		NUMANodeAffinity: affinity,
+		Preferred:        true,
+	}
+}
+
```
I think this is a bit to specific to just this use case with the hard-coding of a single hint with an affinity of `2` deep in the `GetAffinity()` call. If you want to build a `fakeTopologyManager` such as this, we should try and make it a bit more general so that it could be used by other test cases as well.

kikimo replied:
@klueska just update the unit test code, PTAL

kikimo replied:
Another way might be reuse the existing topologymanager.NewFakeManager by passing a fake Hint as its parameter and store in topologymanager.fakeManager{}, but that would require more existing code to be changed, do you think this is a good idea?

kikimo replied:
@klueska I've update to use existing code in:

https://github.com/kubernetes/kubernetes/blob/1295b2c4b59b8bba58186294eb2d76a7ecb4987e/pkg/kubelet/cm/topologymanager/fake_topology_manager.go#L25-L36

fromanirh commented on pkg/kubelet/cm/devicemanager/manager_test.go:
```diff
@@ -669,6 +670,106 @@ type TestResource struct {
+	}
+
+	for _, testCase := range testCases {
+		fromAffinity, notFromAffinity, withoutTopology := testManager.filterByAffinity("", "", "res1", testCase.available)
```
Not a fan of this: in general we should always test public interfaces (and fix the interface accordingly if it is too hard), but in this case it makes sense.

kikimo replied:
aha, I think we can do this by extending code in TestPodContainerDeviceAllocation

fromanirh replied:
If you have the chance please do. Reusing/extending existing testing infra (which should also something @klueska suggested IIUC) and testing against the public interface is something that almost always is a good direction.

kikimo replied:
@fromanirh  I tried but found it too hard to reproduce the problem steadily by testing the public `Allocate()`interface  like in `TestPodContainerDeviceAllocation()`, the problem is that we are using sets.String{} to store the filter results, and the sequence produced by a loop of sets.String{} is random:

https://github.com/kubernetes/kubernetes/blob/7ccd90e7d7c6d727a7497c901c45776592ec5e45/pkg/kubelet/cm/devicemanager/manager.go#L675-L685

also`devicesToAllocate()` will collect result from all the filter result: aligned, unaligned, noAffinity: 

https://github.com/kubernetes/kubernetes/blob/7ccd90e7d7c6d727a7497c901c45776592ec5e45/pkg/kubelet/cm/devicemanager/manager.go#L706-L750

so the result is nondeterministic, I got 74 failed results when I run the single test for 1024 times. In another way, making `filterByAffinity()` a public interface does not seems like a good idea, any better suggestion?

fromanirh replied:
in this case testing the private method, lacking better approaches, is OK with me.
In general I'd like to review later on if we need to use `sets.String` or if a simple ordered `[]string` could suffice and enable better testing. Not blocking for this PR anyway.

kikimo committed:
add a reusable fakeTopologyManagerWithHint{}

kikimo committed:
simplify sorting comparator of numa nodes

kikimo committed:
fix incompatible interface in fakeTopologyManagerWithHint

kikimo commented:
@fromanirh @klueska any update?

klueska commented on pkg/kubelet/cm/devicemanager/manager.go:
```diff
@@ -791,9 +799,27 @@ func (m *ManagerImpl) filterByAffinity(podUID, contName, resource string, availa
 		nodes = append(nodes, node)
 	}
 
-	// Sort the list of nodes by how many devices they contain.
+	getNodePriority := func(n int) nodePriority {
+		if hint.NUMANodeAffinity.IsSet(n) {
+			return priorityFromAffinity
+		}
+
+		if n != nodeWithoutTopology {
+			return priorityNotFromAffinity
+		}
+
+		return priorityWithoutTopology
+	}
+
+	// Sort the list of nodes by nodes from affinity set, nodes from none-affinity
+	// set, nodes without topology set and how many devices they contain.
 	sort.Slice(nodes, func(i, j int) bool {
-		return perNodeDevices[i].Len() < perNodeDevices[j].Len()
+		pi, pj := getNodePriority(nodes[i]), getNodePriority(nodes[j])
+		if pi != pj {
+			return pi < pj
+		}
+
+		return perNodeDevices[nodes[i]].Len() < perNodeDevices[nodes[j]].Len()
```
I was thinking something more like:
```
// Sort the list of nodes by:
// 1) Nodes contained in the 'hint's affinity set
// 2) Nodes not contained in the 'hint's affinity set
// 3) The the fake NUMANode of -1 (assuming it is included in the list)
// Within each of the groups above, sort the nodes by how many devices they contain
sort.Slice(nodes, func(i, j int) bool {
	// If one or the other of nodes[i] or nodes[j] is in the 'hint's affinity set
	if hint.NUMANodeAffinity.IsSet(nodes[i]) && hint.NUMANodeAffinity.IsSet(nodes[j]) {
		return nodes[i] < nodes[j]
	}
	if hint.NUMANodeAffinity.IsSet(nodes[i]) {
		return true
	}
	if hint.NUMANodeAffinity.IsSet(nodes[j]) {
		return false
	}
	
	// If either nodes[i] or nodes[j] are in the 'hint's affinity set (but are not -1)
	if nodes[i] != nodeWithoutTopology && nodes[j] != nodeWithoutTopology {
		return nodes[i] < nodes[j]
	}
	if nodes[i] != nodeWithoutTopology {
		return true
	}
	if nodes[j] != nodeWithoutTopology {
		return false
	}
	
	// If both nodes[i] and nodes[j] are -1
	return nodes[i] < nodes[j]
}
```

kikimo replied:
// Within each of the groups above, sort the nodes by how many devices they contain
should this line of code `return nodes[i] < nodes[j]` be something like `return perNodeDevices[nodes[i]].Len() < perNodeDevices[nodes[j]].Len()`

kikimo committed:
simplify sorting comparator of numa nodes

kikimo commented:
@klueska node sorting algorithm updated, PTAL

klueska commented on pkg/kubelet/cm/devicemanager/manager.go:
```diff
@@ -791,9 +791,36 @@ func (m *ManagerImpl) filterByAffinity(podUID, contName, resource string, availa
+			return false
+		}
+
+		// If either nodes[i] or nodes[j] are in the 'hint's affinity set (but are not -1)
+		if nodes[i] != nodeWithoutTopology && nodes[j] != nodeWithoutTopology {
+			return perNodeDevices[nodes[i]].Len() < perNodeDevices[nodes[j]].Len()
+		}
+		if nodes[i] != nodeWithoutTopology {
+			return true
+		}
+		if nodes[j] != nodeWithoutTopology {
+			return false
+		}
+
+		// If both nodes[i] and nodes[j] are -1
+		return perNodeDevices[nodes[i]].Len() < perNodeDevices[nodes[j]].Len()
```
After seeing this written out, I actually think it would be clearer as:
```
// If one or the other of nodes[i] or nodes[j] is the fake NUMA node -1 (they can't both be)
if nodes[i] == nodeWithoutTopology {
	return false
}
if nodes[j] == nodeWithoutTopology {
	return true
}

// Otherwise both nodes[i] and nodes[j] are real NUMA nodes that are not in the 'hint's' affinity list.
return perNodeDevices[nodes[i]].Len() < perNodeDevices[nodes[j]].Len()
```

Sorry for all the churn.

kikimo replied:
Updated PTAL

klueska replied:
I don't see it updated to this.

klueska replied:
I see now. You added this, but didn't remove the code it was supposed to replace.

kikimo committed:
minor tweak on numa node sorting algorithm

kikimo commented:
> @kikimo: The following test **failed**, say `/retest` to rerun all failed tests:
> 
> Test name	Commit	Details	Rerun command
> pull-kubernetes-e2e-gce-ubuntu-containerd	[445b9c0](https://github.com/kubernetes/kubernetes/commit/445b9c07622793bb7cbcad2b9b062182e940eaee)	[link](https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/101893/pull-kubernetes-e2e-gce-ubuntu-containerd/1395173784423501824/)	`/test pull-kubernetes-e2e-gce-ubuntu-containerd`
> [Full PR test history](https://prow.k8s.io/pr-history?org=kubernetes&repo=kubernetes&pr=101893). [Your PR dashboard](https://prow.k8s.io/pr?query=is%3Apr%20state%3Aopen%20author%3Akikimo). Please help us cut down on flakes by [linking to](https://git.k8s.io/community/contributors/devel/sig-testing/flaky-tests.md#filing-issues-for-flaky-tests) an [open issue](https://github.com/kubernetes/kubernetes/issues?q=is:issue+is:open) when you hit one in your PR.

/retest

kikimo commented:
> link

/retest

kikimo commented:
> @kikimo: The following test **failed**, say `/retest` to rerun all failed tests:
> 
> Test name	Commit	Details	Rerun command
> pull-kubernetes-e2e-gce-ubuntu-containerd	[445b9c0](https://github.com/kubernetes/kubernetes/commit/445b9c07622793bb7cbcad2b9b062182e940eaee)	[link](https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/101893/pull-kubernetes-e2e-gce-ubuntu-containerd/1395193064246480896/)	`/test pull-kubernetes-e2e-gce-ubuntu-containerd`
> [Full PR test history](https://prow.k8s.io/pr-history?org=kubernetes&repo=kubernetes&pr=101893). [Your PR dashboard](https://prow.k8s.io/pr?query=is%3Apr%20state%3Aopen%20author%3Akikimo). Please help us cut down on flakes by [linking to](https://git.k8s.io/community/contributors/devel/sig-testing/flaky-tests.md#filing-issues-for-flaky-tests) an [open issue](https://github.com/kubernetes/kubernetes/issues?q=is:issue+is:open) when you hit one in your PR.

/retest

kikimo commented:
> @kikimo: The following test **failed**, say `/retest` to rerun all failed tests:
> 
> Test name	Commit	Details	Rerun command
> pull-kubernetes-e2e-gce-ubuntu-containerd	[445b9c0](https://github.com/kubernetes/kubernetes/commit/445b9c07622793bb7cbcad2b9b062182e940eaee)	[link](https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/101893/pull-kubernetes-e2e-gce-ubuntu-containerd/1395217293104910336/)	`/test pull-kubernetes-e2e-gce-ubuntu-containerd`
> [Full PR test history](https://prow.k8s.io/pr-history?org=kubernetes&repo=kubernetes&pr=101893). [Your PR dashboard](https://prow.k8s.io/pr?query=is%3Apr%20state%3Aopen%20author%3Akikimo). Please help us cut down on flakes by [linking to](https://git.k8s.io/community/contributors/devel/sig-testing/flaky-tests.md#filing-issues-for-flaky-tests) an [open issue](https://github.com/kubernetes/kubernetes/issues?q=is:issue+is:open) when you hit one in your PR.

/retest

kikimo commented:
@klueska all updated, can we merge now?

fromanirh commented on pkg/kubelet/cm/devicemanager/manager_test.go:
```diff
@@ -753,23 +759,45 @@ func TestFilterByAffinity(t *testing.T) {
+// NewFakeTopologyManagerWithHint returns an instance of fake topology manager with specified topology hints
+func NewFakeTopologyManagerWithHint(t *testing.T, hint *topologymanager.TopologyHint) topologymanager.Manager {
+	return &fakeTopologyManagerWithHint{
+		t:    t,
```
not sure we need the logs, but won't hurt either

ffromani reviewed (commented):
I'd like to spend some time to grok the new algo - today and tomorrow likely; at glance I also like the new direction much more. Added some initial comments, nothing blocking from my side. If @klueska is happy with the changes no need to wait for me, I'll catch up later.

kikimo committed:
remove redundant test branch in sorting algorithm

kikimo commented:
@klueska any further comment?

klueska commented:
/approve
/lgtm
